app/models/user.rb

```diff
@@ -940,6 +940,184 @@ def remove_image(image)
+    delete_unattached_images
+    User.erase_user(id) if no_references_left?
+  end
+
+  # Disable and remove all public information from account but leave it there
+  # in case there are still comments, etc. on the site by this user.
+  def disable_account
+    self.password = ""
+    self.email = ""
+    self.blocked = true
+    self.location_id = nil
+    self.image_id = nil
+    self.notes = nil
+    self.mailing_address = nil
+    self.save
+  end
+
+  def delete_api_keys
+    api_keys.delete_all
+  end
+
+  def delete_interests
+    interests.delete_all
+  end
+
+  def delete_notifications
+    notifications.delete_all
+  end
+
+  def delete_queued_emails
+    QueuedEmail.where(user_id: id).delete_all
+    QueuedEmail.where(to_user_id: id).delete_all
+  end
+
+  def delete_observations
+    [ Naming, Vote, RssLog ].each do |model|
+      model.joins(:observation).where(observation: { user_id: id }).delete_all
+    end
+    # (all the rest of the observations' dependents should autodestruct)
+    observations.delete_all
+  end
+
+  # Delete user's descriptions that don't have any other authors or # editors.
+  def delete_private_name_descriptions
+    (name_descriptions -
+      name_descriptions.joins(:name_description_authors).
+        where.not(name_description_authors: { user_id: id }) -
+      name_descriptions.joins(:name_description_editors).
+        where.not(name_description_editors: { user_id: id }))
+      delete_all
+  end
+
+  # Delete user's descriptions that don't have any other authors or # editors.
+  def delete_private_location_descriptions
+    (location_descriptions -
+      location_descriptions.joins(:location_description_authors).
+        where.not(location_description_authors: { user_id: id }) -
+      location_descriptions.joins(:location_description_editors).
+        where.not(location_description_editors: { user_id: id }))
+      delete_all
+  end
+
+  # Delete all the user's projects that don't have any other users on them.
+  def delete_private_projects
+    (projects.select do |project|
+      project.admin_group_users.
+        where.not(admin_group_users: { user_id: id }).none? &&
+      project.member_group_users.
+        where.not(member_group_users: { user_id: id }).none?
+    end).delete_all
+  end
+
+  # Delete all species lists the user created unless they belong to a project.
+  # (Private projects should already have been deleted by this point, so this
+  # in effect, really should read "unless they belong to a public project".)
+  # I think it's okay to delete observations even if they are attached to a
+  # project.  But species_lists are potentially a much more collaborative
+  # effort, so I don't think it's okay to delete lists that are attached to
+  # public projects just because the user happened to originally create them.
+  # -JPH 20220916
+  def delete_private_species_lists
+    (species_lists -
+      species_lists.joins(:project_species_lists)).
+      delete_all
+  end
```

Comment:
Tried this locally. `delete_all` needs an AR relation. I think it needs to be something like 
```ruby
    ids = (species_lists -
            species_lists.joins(:project_species_lists)).map(&:id)
    SpeciesList.where(id: ids).delete_all
```
This works, but reveals another issue: test fixtures shows 8 SLs for mary, but only 3 where SL.joins(PSL). 

That leaves 3 not deleted by `UserTest#test_delete_private_species_lists`, and the test is only expecting one.